distributed/tests/test_spans.py

```diff
@@ -582,3 +582,34 @@ async def test_merge_by_tags_metrics(c, s, a, b):
     assert ext.merge_by_tags("foo").enqueued == min(
         ext.spans[foo1].enqueued, ext.spans[foo2].enqueued
     )
+
+
+@gen_cluster(client=True, config={"distributed.diagnostics.computations.nframes": 10})
+async def test_code(c, s, a, b):
+    with span("foo") as foo:
+        for subspan in ("bar", "baz"):
+            with span(subspan):
+                for key in ("x", "y"):
+                    # Call update-graph four times in two different subspans,
+                    # all with the same code stack
+                    await c.submit(inc, 1, key=key)
+
+        await c.submit(inc, 2, key="z")
+        await c.submit(inc, 3, key="w")
+
+    code = s.extensions["spans"].spans[foo].code
+
+    # Identical code stacks have been deduplicated
+    assert len(code) == 3
+    assert "def _run(self)" in code[0][-2].code
+    # Code snippets are in first-seen order
+    assert "inc, 1" in code[0][-1].code
+    assert "inc, 2" in code[1][-1].code
+    assert "inc, 3" in code[2][-1].code
```

Comment:
Based on my testing, all spans have the same code. This is not a blocker but a bit unforunate.

I played briefly with `ast` to see if we can improve this. This code is a bit of a hack but works and we could apply something like this on a best-effort basis to only attach the code that is relevant to the given span.

I'm basing this on the `span` function name in the contextmanager. This could be refined further but should be good enough for most cases.

```python
import ast
def parse_frame_by_span(frame):
    code_by_span = {}
    ast_fr = ast.parse(inspect.getsource(fr))
    for node in ast.walk(ast_fr):
        if isinstance(node, ast.With) and node.items[0].context_expr.func.id == "span":
            try:
                # This is a literal
                label = node.items[0].context_expr.args[0].value
            except:
                # This is a variable. We can't get all of them, only whatever
                # is stored in the locals at the time. Still, nice.
                label = node.items[0].context_expr.args[0].id
                label = fr.f_locals[label]
            code_by_span[label] = ast.unparse(node)
    return code_by_span
```

![image](https://github.com/dask/distributed/assets/8629629/65303229-f4a9-4ed8-98c8-529906ec1d64)